src/equations/shallow_water_2d.jl

```diff
@@ -199,6 +197,71 @@ For details see Section 9.2.5 of the book:
   return flux
 end
 
+"""
+    boundary_condition_slip_wall(u_inner, normal_direction, direction, x, t,
+                                 surface_flux_function, equations::ShallowWaterEquations2D)
+
+Should be used together with [`StructuredMesh`](@ref).
+"""
+@inline function boundary_condition_slip_wall(u_inner, normal_direction::AbstractVector,
+                                              direction, x, t,
+                                              surface_flux_function,
+                                              equations::ShallowWaterEquations2D)
+  # flip sign of normal to make it outward pointing, then flip the sign of the normal flux back
+  # to be inward pointing on the -x and -y sides due to the orientation convention used by StructuredMesh
+  if isodd(direction)
+    boundary_flux = -boundary_condition_slip_wall(u_inner, -normal_direction,
+                                                  x, t, surface_flux_function, equations)
+  else
+    boundary_flux = boundary_condition_slip_wall(u_inner, normal_direction,
+                                                 x, t, surface_flux_function, equations)
+  end
+
+  return boundary_flux
+end
+
+"""
+    boundary_condition_slip_wall(u_inner, orientation, direction, x, t,
+                                 surface_flux_function, equations::ShallowWaterEquations2D)
+
+Should be used together with [`TreeMesh`](@ref).
+"""
+@inline function boundary_condition_slip_wall(u_inner, orientation,
+                                              direction, x, t,
+                                              surface_flux_function,
+                                              equations::ShallowWaterEquations2D)
+  # get the appropriate normal vector from the orientation
+  if orientation == 1
+    if isodd(direction)
+      normal_direction = SVector(-1, 0)
+    else
+      normal_direction = SVector(1, 0)
+    end
+  else # orientation == 2
+    if isodd(direction)
+      normal_direction = SVector(0, -1)
+    else
+      normal_direction = SVector(0, 1)
+    end
+  end
+
+  # normalize the outward pointing direction
+  normal = normal_direction / norm(normal_direction)
+
+  # compute the normal velocity
+  u_normal = normal[1] * u_inner[2] + normal[2] * u_inner[3]
+
+  # create the "external" boundary solution state
+  u_boundary = SVector(u_inner[1],
+                       u_inner[2] - 2.0 * u_normal * normal[1],
+                       u_inner[3] - 2.0 * u_normal * normal[2],
+                       u_inner[4])
+
+  # compute and return the flux using `boundary_condition_slip_wall` routine above
+  flux = surface_flux_function(u_inner, u_boundary, orientation, equations)
```

Comment:
On the `TreeMesh` we can make this simpler because we do not need to account for the `direction`. This can shrink to become
```julia
  # get the appropriate normal vector from the orientation
  if orientation == 1
    u_boundary = SVector(u_inner[1], -u_inner[2], u_inner[3], u_inner[4])
  else # orientation == 2
    u_boundary = SVector(u_inner[1], u_inner[2], -u_inner[3], u_inner[4])
  end

  # compute and return the flux using `boundary_condition_slip_wall` routine above
  flux = surface_flux_function(u_inner, u_boundary, orientation, equations)
```

Reply:
Considered in latest  commit